src/Compiler/Driver/OptimizeInputs.fs

```diff
@@ -74,131 +326,219 @@ let ApplyAllOptimizations
-                            tcConfig.emitTailcalls,
-                            hidden,
-                            implFile
-                        )
-
-                    //PrintWholeAssemblyImplementation tcConfig outfile (sprintf "extra-loop-%d" n) implFile
-                    implFile, optEnvExtraLoop
-                else
-                    implFile, optEnvExtraLoop
-
-            let implFile =
-                if tcConfig.doDetuple then
-                    //ReportTime tcConfig ("Detupled optimization")
-                    let implFile = implFile |> Detuple.DetupleImplFile ccu tcGlobals
-                    //PrintWholeAssemblyImplementation tcConfig outfile "post-detuple" implFile
-                    implFile
-                else
-                    implFile
-
-            let implFile =
-                if tcConfig.doTLR then
-                    implFile
-                    |> InnerLambdasToTopLevelFuncs.MakeTopLevelRepresentationDecisions ccu tcGlobals
-                else
-                    implFile
-
-            let implFile = LowerCalls.LowerImplFile tcGlobals implFile
-
-            let implFile, optEnvFinalSimplify =
-                if tcConfig.doFinalSimplify then
-
-                    //ReportTime tcConfig ("Final simplify pass")
-                    let (optEnvFinalSimplify, implFile, _, _), _ =
-                        Optimizer.OptimizeImplFile(
-                            optSettings,
-                            ccu,
-                            tcGlobals,
-                            tcVal,
-                            importMap,
-                            optEnvFinalSimplify,
-                            isIncrementalFragment,
-                            tcConfig.fsiMultiAssemblyEmit,
-                            tcConfig.emitTailcalls,
-                            hidden,
-                            implFile
-                        )
-
-                    //PrintWholeAssemblyImplementation tcConfig outfile "post-rec-opt" implFile
-                    implFile, optEnvFinalSimplify
-                else
-                    implFile, optEnvFinalSimplify
-
-            let implFile =
+    let addPhaseDiagnostics (f: PhaseFunc) (info: Phase) =
+        fun (inputs: PhaseInputs) ->
+            use _ =
+                let tags =
+                    [|
+                        "QualifiedNameOfFile", inputs.File.QualifiedNameOfFile.Text
+                        "OptimisationPhase", info.Name
+                    |]
+
+                FSharp.Compiler.Diagnostics.Activity.start $"file-{inputs.FileIdx}_phase-{info.Name}" tags
+
+            f inputs
+
+    let phases = List<PhaseInfo>()
+
+    let addPhase (name: string) (phaseFunc: PhaseFunc) =
+        let phase = { Idx = phases.Count; Name = name }
+
+        let phaseInfo =
+            {
+                Phase = phase
+                Func = addPhaseDiagnostics phaseFunc phase
+            }
+
+        phases.Add(phaseInfo)
+
+    let firstLoop
```

Comment:
These below functions are/should be equivalent to the code originally in the `List.fold`, but divided into phases so that parallel optimisation can handle them separately. I didn't intend to change the behaviour of sequential optimization.